validator/validator/metrics.py

```diff
@@ -1,65 +1,71 @@
-    watts_used: list[float]
+    metrics: list[MetricData]
 
     def __init__(self, metagraph: bt.metagraph):
         self.metagraph = metagraph
         self.clear()
 
     def clear(self):
-        self.baseline_averages = [0.0] * self.metagraph.n.item()
-        self.model_averages = [0.0] * self.metagraph.n.item()
-        self.similarity_averages = [0.0] * self.metagraph.n.item()
-        self.sizes = [0] * self.metagraph.n.item()
-        self.vram_used = [0.0] * self.metagraph.n.item()
-        self.watts_used = [0.0] * self.metagraph.n.item()
+        self.metrics = [new_metric_data()] * self.metagraph.n.item()
 
     def reset(self, uid: int):
-        self.baseline_averages[uid] = 0.0
-        self.model_averages[uid] = 0.0
-        self.similarity_averages[uid] = 0.0
-        self.sizes[uid] = 0
-        self.vram_used[uid] = 0.0
-        self.watts_used[uid] = 0.0
+        self.metrics[uid] = new_metric_data()
 
     def update(self, uid: int, benchmark: CheckpointBenchmark):
-        self.baseline_averages[uid] = benchmark.baseline_average
-        self.model_averages[uid] = benchmark.average_time
-        self.similarity_averages[uid] = benchmark.average_similarity
-        self.sizes[uid] = benchmark.size
-        self.vram_used[uid] = benchmark.vram_used
-        self.watts_used[uid] = benchmark.watts_used
+        self.metrics[uid] = MetricData(
+            baseline_average=benchmark.baseline_average,
+            model_average=benchmark.average_time,
+            similarity_average=benchmark.average_similarity,
+            size=benchmark.size,
+            vram_used=benchmark.vram_used,
+            watts_used=benchmark.watts_used
+        )
 
     def resize(self):
-        def resize_data(data: list) -> list:
-            new_data = [0.0] * self.metagraph.n.item()
-            length = len(self.metagraph.hotkeys)
-            new_data[:length] = data[:length]
-            return new_data
-
-        self.baseline_averages = resize_data(self.baseline_averages)
-        self.model_averages = resize_data(self.model_averages)
-        self.similarity_averages = resize_data(self.similarity_averages)
-        self.sizes = resize_data(self.sizes)
-        self.vram_used = resize_data(self.vram_used)
-        self.watts_used = resize_data(self.watts_used)
-
-    def calculate_score(self, uid: int) -> float:
-        return max(
-            0.0,
-            self.baseline_averages[uid] - self.model_averages[uid]
-        ) * self.similarity_averages[uid]
+        new_data = [new_metric_data()] * self.metagraph.n.item()
+        length = len(self.metagraph.hotkeys)
+        new_data[:length] = self.metrics[:length]
+        self.metrics = new_data
+
+    def get_sorted_contestants(self) -> list[tuple[Uid, float]]:
+        contestants = []
+        for uid in range(self.metagraph.n.item()):
+            metric_data = self.metrics[uid]
+            contestants.append((Uid(uid), metric_data.calculate_score()))
```

Comment:
doesn't `Uid(uid)` cause a type error at runtime?

Reply:
pycharm was complaining about types without it. but I gave it a test and it seems to work fine without the UID type. this was also causing a circular import anyway